### PR TITLE
Ft gan trainer

### PR DESCRIPTION
**Pull request by MutegekiHenry**
# Summary

## Add support for distributed GAN model training using pytorch and MNIST. 

This PR adds functionality to distribute the training of a GAN model in a distributed GPU enviroment by extending the TorchTrainer Class to cater for the Discriminator and Generator training. The following were added through this work:

- The tutorial working directory is `tutorials/distributed-ml/torch-tutorial-GAN`
- Added a simple GAN example using the MNIST dataset that serves as a baseline model and works non distributed, refer to the `simpleGAN.py` file. 
- Added a `GANTrainer` that extends the TorchTrainer class and overrides the `train_epoch` , `train_step` , `validation_epoch` and `validation_step` while also creating custom optimizers for the Generator and Discriminator models. This is found in the `train.py` file. 
- Added  `slurm.sh` to ease submssion of multiple jobs via two distributed strategies `dpp` & `deepspeed`. These can be run at once using the `runall.sh` script. 
- Added a readME.md to document the run steps and give context on the distribution implementation. 

---

**Related issue :**

### CODE REVIEW

**Comment by matbun** [comment]
@VerderK could you add this tutorial under the pytorch tutorials please?

File: tutorials/distributed-ml/torch-tutorial-GAN/train.py [hunk 1]
```diff
@@ -0,0 +1,471 @@
...
+)
+
+REAL_LABEL = 1
+FAKE_LABEL = 0
+Z_DIM = 100
+G_HIDDEN = 64
+IMAGE_CHANNEL = 1
+D_HIDDEN = 64
+
+NOISE_DIM = 100
```
Comment (matbun) [inline tutorials/distributed-ml/torch-tutorial-GAN/train.py:30]: these need to be moved to the trainer config and/or Generator/Discrimitator constructor

File: tutorials/distributed-ml/torch-tutorial-GAN/train.py [hunk 2]
```diff
@@ -0,0 +1,471 @@
...
+
+    def forward(self, input):
+        return self.main(input).view(-1, 1).squeeze(1)
+
+
+class GANTrainer(TorchTrainer):
+    def __init__(
+            self,
+            config: Union[Dict, TrainingConfiguration],
+            epochs: int, discriminator: nn.Module,
```
Comment (matbun) [inline tutorials/distributed-ml/torch-tutorial-GAN/train.py:96]: missing new line after ...`int,`

File: tutorials/distributed-ml/torch-tutorial-GAN/train.py [hunk 3]
```diff
@@ -0,0 +1,471 @@
...
+        g_train_losses = []
+        d_train_losses = []
+        for batch_idx, (real_images, _) in enumerate(self.train_dataloader):
+            lossG, lossD = self.train_step(
+                real_images, batch_idx)
+            g_train_losses.append(lossG)
+            d_train_losses.append(lossD)
+            self.log(
+                item=lossG,
+                identifier='generator training loss per epoch',
```
Comment (matbun) [inline tutorials/distributed-ml/torch-tutorial-GAN/train.py:171]: `identifier='generator training loss per batch'`

File: tutorials/distributed-ml/torch-tutorial-GAN/train.py [hunk 4]
```diff
@@ -0,0 +1,471 @@
...
+                step=self.train_glob_step,
+                batch_idx=batch_idx
+            )
+
+            self.train_glob_step += 1
+        # Aggregate and log losses
+        avg_g_loss = torch.mean(torch.stack(g_train_losses))
+        self.log(
+            item=avg_g_loss.item(),
+            identifier='g_train_loss_av per batch',
```
Comment (matbun) [inline tutorials/distributed-ml/torch-tutorial-GAN/train.py:189]: `identifier='generator training loss per epoch'`

File: tutorials/distributed-ml/torch-tutorial-GAN/train.py [hunk 5]
```diff
@@ -0,0 +1,471 @@
...
+            kind='metric',
+            step=epoch,
+        )
+
+        avg_d_loss = torch.mean(torch.stack(d_train_losses))
+        self.log(
+            item=avg_d_loss.item(),
+            identifier='d_train_loss_av per batch',
```
Comment (matbun) [inline tutorials/distributed-ml/torch-tutorial-GAN/train.py:197]: `identifier='discriminator training loss per epoch'`

File: tutorials/distributed-ml/torch-tutorial-GAN/train.py [hunk 6]
```diff
@@ -0,0 +1,471 @@
...
+        )
+
+        self.save_plots_and_images(epoch)
+
+    def validation_epoch(self, epoch: int):
+        g_validation_losses = []
+        g_validation_accuracy = []
+        self.discriminator.eval()
+        self.generator.eval()
+        validation_loss = torch.tensor(0.0, device=self.device)
```
Comment (matbun) [inline tutorials/distributed-ml/torch-tutorial-GAN/train.py:209]: not needed, remove

File: tutorials/distributed-ml/torch-tutorial-GAN/train.py [hunk 7]
```diff
@@ -0,0 +1,471 @@
...
+
+    def train_step(self, real_images, batch_idx):
+        real_images = real_images.to(self.device)
+        batch_size = real_images.size(0)
+        real_labels = torch.full(
+            (batch_size,), REAL_LABEL,
+            dtype=torch.float, device=self.device)
+        fake_labels = torch.full(
+            (batch_size,), FAKE_LABEL,
+            dtype=torch.float, device=self.device)
```
Comment (matbun) [inline tutorials/distributed-ml/torch-tutorial-GAN/train.py:244]: use torch zeros and ones, and remove `REAL_LABEL` and `FAKE_LABEL` from this file.

File: tutorials/distributed-ml/torch-tutorial-GAN/train.py [hunk 8]
```diff
@@ -0,0 +1,471 @@
...
+
+        return lossG, loss
+
+    def validation_step(self, real_images, batch_idx):
+        real_images = real_images.to(self.device)
+        batch_size = real_images.size(0)
+        real_labels = torch.full((batch_size,), REAL_LABEL,
+                                 dtype=torch.float, device=self.device)
+        fake_labels = torch.full((batch_size,), FAKE_LABEL,
+                                 dtype=torch.float, device=self.device)
```
Comment (matbun) [inline tutorials/distributed-ml/torch-tutorial-GAN/train.py:277]: use torch zeros and ones, and remove `REAL_LABEL` and `FAKE_LABEL` from this file.

File: tutorials/distributed-ml/torch-tutorial-GAN/train.py [hunk 9]
```diff
@@ -0,0 +1,471 @@
...
+            dtype=torch.float, device=self.device)
+        fake_labels = torch.full(
+            (batch_size,), FAKE_LABEL,
+            dtype=torch.float, device=self.device)
+
+        # Train Discriminator with real images
+        output_real = self.discriminator(real_images)
+        lossD_real = self.criterion(output_real, real_labels)
+        # Generate fake images and train Discriminator
+        noise = torch.randn(batch_size, NOISE_DIM, 1, 1, device=self.device)
```
Comment (matbun) [inline tutorials/distributed-ml/torch-tutorial-GAN/train.py:250]: `NOISE_DIM` should be received from `self.config`. Remove `NOISE_DIM` from this file.

File: tutorials/distributed-ml/torch-tutorial-GAN/train.py [hunk 10]
```diff
@@ -0,0 +1,471 @@
...
+                                 dtype=torch.float, device=self.device)
+        fake_labels = torch.full((batch_size,), FAKE_LABEL,
+                                 dtype=torch.float, device=self.device)
+
+        # Validate with real images
+        output_real = self.discriminator(real_images)
+        loss_real = self.criterion(output_real, real_labels)
+
+        # Generate and validate fake images
+        noise = torch.randn(batch_size, NOISE_DIM, 1, 1, device=self.device)
```
Comment (matbun) [inline tutorials/distributed-ml/torch-tutorial-GAN/train.py:284]: `NOISE_DIM` should be received from `self.config`. Remove `NOISE_DIM` from this file.

File: tutorials/distributed-ml/torch-tutorial-GAN/train.py [hunk 11]
```diff
@@ -0,0 +1,471 @@
...
+        real_images = real_images.to(self.device)
+        batch_size = real_images.size(0)
+        real_labels = torch.full((batch_size,), REAL_LABEL,
+                                 dtype=torch.float, device=self.device)
+        fake_labels = torch.full((batch_size,), FAKE_LABEL,
+                                 dtype=torch.float, device=self.device)
+
+        # Validate with real images
+        output_real = self.discriminator(real_images)
+        loss_real = self.criterion(output_real, real_labels)
```
Comment (matbun) [inline tutorials/distributed-ml/torch-tutorial-GAN/train.py:281]: compute also accuracy of disc on real images

File: tutorials/distributed-ml/torch-tutorial-GAN/train.py [hunk 12]
```diff
@@ -0,0 +1,471 @@
...
+        loss_fake = self.criterion(output_fake, fake_labels)
+
+        # Generator's attempt to fool the discriminator
+        loss_gen = self.criterion(output_fake, real_labels)
+        accuracy_gen = (
+            (output_fake > 0.5).float() == real_labels).float().mean()
+
+        # Calculate total discriminator loss and accuracy
+        d_total_loss = (loss_real + loss_fake) / 2
+        accuracy = ((output_real > 0.4).float() == real_labels).float().mean(
```
Comment (matbun) [inline tutorials/distributed-ml/torch-tutorial-GAN/train.py:298]: why 0.4 and not 0.5?

File: tutorials/distributed-ml/torch-tutorial-GAN/train.py [hunk 13]
```diff
@@ -0,0 +1,471 @@
...
+            item=d_accuracy,
+            identifier='discriminator validation_accuracy',
+            kind='metric',
+            step=self.validation_glob_step,
+            batch_idx=batch_idx
+        )
+        return loss_gen, accuracy_gen
+
+    def configure_optimizers(self):
+        return [self.optimizerD, self.optimizerG]
```
Comment (matbun) [inline tutorials/distributed-ml/torch-tutorial-GAN/train.py:334]: this method is not needed... right?

File: tutorials/distributed-ml/torch-tutorial-GAN/train.py [hunk 14]
```diff
@@ -0,0 +1,471 @@
...
+
+        if 'lr_scheduler' in checkpoint:
+            if checkpoint['lr_scheduler'] is not None:
+                self.lr_scheduler.load_state_dict(checkpoint['lr_scheduler'])
+
+        print(f"Checkpoint loaded from {checkpoint_path}")
+
+    def save_plots_and_images(self, epoch):
+        self.generator.eval()
+        noise = torch.randn(64, NOISE_DIM, 1, 1, device=self.device)
```
Comment (matbun) [inline tutorials/distributed-ml/torch-tutorial-GAN/train.py:374]: Same comment on `NOISE_DIM` as above

File: tutorials/distributed-ml/torch-tutorial-GAN/train.py [hunk 15]
```diff
@@ -0,0 +1,471 @@
...
+    def weights_init(m):
+        classname = m.__class__.__name__
+        if classname.find('Conv') != -1:
+            m.weight.data.normal_(0.0, 0.02)
+        elif classname.find('BatchNorm') != -1:
+            m.weight.data.normal_(1.0, 0.02)
+            m.bias.data.fill_(0)
+
+    # Models
+    netG = Generator().to(torch.device("cuda" if torch.cuda.is_available()
```
Comment (matbun) [inline tutorials/distributed-ml/torch-tutorial-GAN/train.py:430]: you don't need to move the models to  some specific device. The trainer will do it for you. Please remove

File: tutorials/distributed-ml/torch-tutorial-GAN/train.py [hunk 16]
```diff
@@ -0,0 +1,471 @@
...
+                                       else "cpu "))
+    netG.apply(weights_init)
+    netD = Discriminator().to(torch.device("cuda" if torch.cuda.is_available()
```
Comment (matbun) [inline tutorials/distributed-ml/torch-tutorial-GAN/train.py:433]: you don't need to move the models to  some specific device. The trainer will do it for you. Please remove

File: tutorials/distributed-ml/torch-tutorial-GAN/train.py [hunk 17]
```diff
@@ -0,0 +1,471 @@
...
+        loss='cross_entropy'
+    )
+
+    # Logger
+    logger = MLFlowLogger(experiment_name='Distributed GAN MNIST', log_freq=10)
+
+    metrics = {
+        'accuracy': torchmetrics.Accuracy(task='multiclass', num_classes=10),
+        'precision': torchmetrics.Precision(task='multiclass', num_classes=10)
+    }
```
Comment (matbun) [inline tutorials/distributed-ml/torch-tutorial-GAN/train.py:451]: you can actually use these metrics to compute the discrimintator precision, or just remove this dict as it is not used

**Review by matbun — CHANGES_REQUESTED** [review]
Please copy the `GANTrainer` below the `TorchTrainer` under `itwinai.torch.trainer`, so that it is also part of the main library

File: tutorials/distributed-ml/torch-tutorial-GAN/train.py (hunk 13 above)
Comment (MutegekiHenry) [reply tutorials/distributed-ml/torch-tutorial-GAN/train.py:334]: Oh yes it's redundant

File: tutorials/distributed-ml/torch-tutorial-GAN/train.py (hunk 17 above)
Comment (MutegekiHenry) [reply tutorials/distributed-ml/torch-tutorial-GAN/train.py:451]: Oh yes its not being used, let me remove it.

File: src/itwinai/torch/trainer.py [hunk 18]
```diff
@@ -745,6 +758,340 @@ def test_step(
...
+            batch_idx=batch_idx
+        )
+        self.log(
+            item=d_accuracy,
+            identifier='discriminator validation_accuracy',
+            kind='metric',
+            step=self.validation_glob_step,
+            batch_idx=batch_idx
+        )
+        return loss_gen, accuracy_gen
```
Comment (matbun) [inline src/itwinai/torch/trainer.py:1033]: also return and log discriminator metrics

File: src/itwinai/torch/trainer.py [hunk 19]
```diff
@@ -745,6 +758,340 @@ def test_step(
...
+            step=self.validation_glob_step,
+            batch_idx=batch_idx
+        )
+        self.log(
+            item=d_accuracy,
+            identifier='discriminator validation_accuracy',
+            kind='metric',
+            step=self.validation_glob_step,
+            batch_idx=batch_idx
+        )
```
Comment (matbun) [inline src/itwinai/torch/trainer.py:1032]: identifiers should be consistent with training metrics

File: tutorials/distributed-ml/torch-tutorial-GAN/train.py [hunk 20]
```diff
@@ -0,0 +1,455 @@
...
+            # output layer
+            nn.Conv2d(d_hidden * 8, 1, 4, 1, 0, bias=False),
+            nn.Sigmoid()
+        )
+
+    def forward(self, input):
+        return self.main(input).view(-1, 1).squeeze(1)
+
+
+class GANTrainer(TorchTrainer):
```
Comment (matbun) [inline tutorials/distributed-ml/torch-tutorial-GAN/train.py:83]: same comments for GANTrainer above

**Review by matbun — CHANGES_REQUESTED** [review]

File: src/itwinai/torch/trainer.py [hunk 21]
```diff
@@ -745,6 +758,340 @@ def test_step(
...
+            identifier='gen_validation_accuracy_epoch',
+            kind='metric',
+            step=self.validation_glob_step,
+        )
+
+        return validation_loss
+
+    def train_step(self, real_images, batch_idx):
+        real_images = real_images.to(self.device)
+        batch_size = real_images.size(0)
```
[the remark this answers is not in the document]
Comment (MutegekiHenry) [reply src/itwinai/torch/trainer.py:938]: I have read more as to why `real_images.size(0)` is a more efficient way to get the batch size compared to getting it statically from the config. This is because it dynamically computes the size of the image batch which might vary on the last batch since the total number of data points might not always be equally divided in all the batches so the last batch could have less images hence the error that was being raised when I use the static batch_size from the config.

File: src/itwinai/torch/trainer.py [hunk 22]
```diff
@@ -745,6 +758,340 @@ def test_step(
...
+        lossG = self.criterion(output_fake, real_labels)
+        self.optimizerG.zero_grad()
+        lossG.backward()
+        self.optimizerG.step()
+
+        return lossG, loss
+
+    def validation_step(self, real_images, batch_idx):
+        real_images = real_images.to(self.device)
+        batch_size = real_images.size(0)
```
[the remark this answers is not in the document]
Comment (MutegekiHenry) [reply src/itwinai/torch/trainer.py:974]: based on the findings, generating the batch_size dynamically might be more efficient.

File: src/itwinai/torch/trainer.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (matbun) [reply src/itwinai/torch/trainer.py:946]: Good point! You're right. Let's leave it as is. Thanks for checking

File: tutorials/distributed-ml/torch-tutorial-GAN/train.py [hunk 23]
```diff
@@ -0,0 +1,525 @@
...
+    """Trainer class for GAN model using pytorch and MNIST dataset.
```
Comment (matbun) [inline tutorials/distributed-ml/torch-tutorial-GAN/train.py:83]: avoid mentioning MNIST. This trainer is actually more generic that that.

File: src/itwinai/torch/trainer.py [hunk 24]
```diff
@@ -745,6 +758,381 @@ def test_step(
         raise NotImplementedError()
 
 
+class GANTrainer(TorchTrainer):
+    """Trainer class for GAN models using pytorch and MNIST dataset.
```
Comment (matbun) [inline src/itwinai/torch/trainer.py:762]: avoid mentioning MNIST. This trainer is actually more generic that that.

File: tutorials/distributed-ml/torch-tutorial-GAN/train.py [hunk 25]
```diff
@@ -0,0 +1,525 @@
...
+    netG.apply(weights_init)
+    netD = Discriminator(d_hidden=64, image_channel=1)
+    netD.apply(weights_init)
+
+    # Training configuration
+    training_config = TrainingConfiguration(
+        batch_size=args.batch_size,
+        lr=args.lr,
+        epochs=args.epochs,
+        loss='cross_entropy',
```
Comment (matbun) [inline tutorials/distributed-ml/torch-tutorial-GAN/train.py:501]: I think that loss is not needed here

**Review by matbun — CHANGES_REQUESTED** [review]

**Review by matbun — DISMISSED** [review]
